src/devices/sound/stt_sa1.cpp

```diff
@@ -0,0 +1,210 @@
+*/
+#include "emu.h"
+#include "multibyte.h"
+#include "stt_sa1.h"
+
+// #define VERBOSE (LOG_GENERAL)
+// #define LOG_OUTPUT_STREAM std::cout
+#include "logmacro.h"
+
+
+DEFINE_DEVICE_TYPE(STT_SA1, stt_sa1_device, "stt_sa1", "ST-Techno STT-SA1 Sound")
+
+stt_sa1_device::stt_sa1_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, STT_SA1, tag, owner, clock)
+	, device_sound_interface(mconfig, *this)
+	, m_stream(nullptr)
+	, m_readsample_cb(*this, 0)
+{
+}
+
+void stt_sa1_device::enable_w(uint16_t data)
+{
+	m_enabled = data != 0;
+}
+
+uint16_t stt_sa1_device::read(offs_t offset, uint16_t mem_mask)
+{
+	if (offset < 0x100 / 2)
+		return m_regs[offset]; // TODO: Should this return addr_cur for regs 1 and 2 instead?
+
+	return get_u16be(&m_ram[offset * 2]);
+}
+
+void stt_sa1_device::write(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	if (offset < 0x100 / 2)
+	{
+		const int v = offset >> 4;
+		const int reg = offset & 0xf;
+
+		m_regs[offset] = data;
+
+		switch (reg)
+		{
+			case 1:
+				m_voice[v].addr_start = m_voice[v].addr_cur = (m_voice[v].addr_cur & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: start = %08llx\n", v, m_voice[v].addr_cur >> 12);
+				break;
+			case 2:
+				m_voice[v].addr_start = m_voice[v].addr_cur = (m_voice[v].addr_cur & 0x0000ffff000) | (uint64_t(data & 0x7fff) << 28);
+				m_voice[v].is_looped = BIT(data, 15) != 0;
+				LOG("voice %d: start = %08llx, is_looped = %d\n", v, m_voice[v].addr_cur >> 12, m_voice[v].is_looped);
+				break;
+
+			case 3:
+				m_voice[v].freq = data;
+				LOG("voice %d: step = %08x\n", v, m_voice[v].freq);
+				break;
+
+			case 4:
+				m_voice[v].addr_start = (m_voice[v].addr_start & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: addr_start = %08llx\n", v, m_voice[v].addr_start >> 12);
+				break;
+			case 5:
+				m_voice[v].addr_start = (m_voice[v].addr_start & 0x0000ffff000) | (uint64_t(data) << 28);
+				LOG("voice %d: addr_start = %08llx\n", v, m_voice[v].addr_start >> 12);
+				break;
+
+			case 6:
+				m_voice[v].addr_end = (m_voice[v].addr_end & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: addr_end = %08llx\n", v, m_voice[v].addr_end >> 12);
+				break;
+			case 7:
+				m_voice[v].addr_end = (m_voice[v].addr_end & 0x0000ffff000) | (uint64_t(data) << 28);
+				LOG("voice %d: addr_end = %08llx\n", v, m_voice[v].addr_end >> 12);
+				break;
+
+			case 0x0b:
+				m_voice[v].vol_l = data;
+				LOG("voice %d: vol_l = %08x\n", v, m_voice[v].vol_l);
+				break;
+			case 0x0c:
+				m_voice[v].vol_r = data;
+				LOG("voice %d: vol_r = %08x\n", v, m_voice[v].vol_r);
+				break;
+
+			default:
+				LOG("Unknown register usage: voice %d, register %x, data %04x\n", v, reg, data);
+				break;
+		}
+	}
+	else
+	{
+		put_u16be(&m_ram[offset * 2], data);
+	}
+}
+
+uint16_t stt_sa1_device::key_r()
+{
+	if (!m_enabled)
+		return 0;
+
+	m_stream->update();
+	return m_keyctrl;
```

Comment:
Is it possible to get here from a memory read handler?  Does it need a check for disabled side effects before updating the stream?